fe/fe-core/src/main/java/com/starrocks/clone/DynamicPartitionScheduler.java

```diff
@@ -372,6 +387,91 @@ private void executeDynamicPartition() {
+                continue;
+            }
+            Table table = db.getTable(tableId);
+            OlapTable olapTable;
+            if (table instanceof OlapTable) {
+                olapTable = (OlapTable) table;
+            } else {
+                iterator.remove();
+                LOG.warn("Could not get database={}, table={}. remove it from scheduler",
+                        dbId, tableId);
+                continue;
+            }
+
+            PartitionInfo partitionInfo = olapTable.getPartitionInfo();
+            RangePartitionInfo rangePartitionInfo;
+            if (partitionInfo instanceof RangePartitionInfo) {
+                rangePartitionInfo = (RangePartitionInfo) olapTable.getPartitionInfo();
+            } else {
+                LOG.warn("currently only support range partition." +
+                        "remove database={}, table={} from scheduler", dbId, tableId);
+                continue;
+            }
+
+            if (rangePartitionInfo.getPartitionColumns().size() != 1) {
+                iterator.remove();
+                LOG.warn("currently only support partition with single column. " +
+                        "remove database={}, table={} from scheduler", dbId, tableId);
+                continue;
+            }
+
+            int ttlNumber = olapTable.getTableProperty().getPartitionTTLNumber();
+            if (Objects.equals(ttlNumber, NO_TTL)) {
+                iterator.remove();
+                LOG.warn("database={}, table={} have no ttl. remove it from scheduler", dbId, tableId);
+                continue;
+            }
+
+            ArrayList<DropPartitionClause> dropPartitionClauses = getDropPartitionClauseByTTL(olapTable, ttlNumber);
+            String tableName = olapTable.getName();
+            for (DropPartitionClause dropPartitionClause : dropPartitionClauses) {
+                db.writeLock();
+                try {
+                    GlobalStateMgr.getCurrentState().dropPartition(db, olapTable, dropPartitionClause);
+                    clearDropPartitionFailedMsg(tableName);
+                } catch (DdlException e) {
+                    recordDropPartitionFailedMsg(db.getOriginName(), tableName, e.getMessage());
+                } finally {
+                    db.writeUnlock();
+                }
+            }
+
+        }
+    }
+
+    private ArrayList<DropPartitionClause> getDropPartitionClauseByTTL(OlapTable olapTable, int ttlNumber) {
+        ArrayList<DropPartitionClause> dropPartitionClauses = new ArrayList<>();
+        RangePartitionInfo info = (RangePartitionInfo) (olapTable.getPartitionInfo());
+
+        List<Map.Entry<Long, Range<PartitionKey>>> idToRanges = new ArrayList<>(info.getIdToRange(false).entrySet());
+        idToRanges.sort(Comparator.comparing(o -> o.getValue().upperEndpoint()));
+
+        int allPartitionNumber = idToRanges.size();
+        if (allPartitionNumber <= ttlNumber) {
```

Comment:
so the partition_ttl_number doesn't means we drop the partition based on creation time but on the total number of partitions we should keep?

Reply:
yes